### PR TITLE
Late day usage table

### LINKED ISSUE #758: Fix incorrectly displayed late days in Student Information panel
For at least one student late days are displayed incorrectly. Also, status is displayed incorrectly in some instances. See image for example. Student has not used 134 late days. No late days were used on this assignment but it is still marked as bad. 
incorrect_late_days
- [x] Display late days used for each assignment up to and include the assignment being viewed. Do not include any later assignments.
- [x]  Total late and display late days 
- [x] Fix incorrect assignment status
- [x] Give a strong visual indicator that the assignment is bad (red background)

### PR DESCRIPTION
**Pull request by JerameyATyler**
Closes #758 

Added late day usage table to student information section of ta grading. Minor tweaks to TA UI.

### CODE REVIEW

**Comment by bmcutler** [comment]
tweaks:

allowed per term should not subtract the number used.  allowed per term should look in the allowed lates table or the default value from the config.ini.

i'm seeing multiple rows for each assignment.  possibly each submission version.  Maybe we should explicitly print the active version in this table?  (for debugging)

late days charged should be non negative.  If late days used = 0 and extension > 0, late days charged is negative, and remaining days incorrectly increases in value

File: TAGradingServer/account/account-rubric.php [hunk 1]
```diff
@@ -5,11 +5,12 @@
     $s_user_id
     $g_id
 */
-
+require "../toolbox/late-day-calculation.php";
```
Comment (MasterOdin) [inline TAGradingServer/account/account-rubric.php]: Move this to models and do "use \models\LateDayCalculations".

File: TAGradingServer/account/submit/admin-grade-summaries.php [hunk 2]
```diff
@@ -163,24 +151,42 @@ function autogradingTotalAwarded($g_id, $student_id, $active_version){
...
         $this_g["id"] = $gradeable['g_id'];
         $this_g["name"] =  $gradeable['g_title'];
+
+	// TODO: DEPRECATE THIS FIELD
         $this_g["score"] = max(0,(floatval($gradeable['score'])+floatval($autograding_score)));
 
-        // adds late days for electronic gradeables 
+	// REPLACED BY:
+        $this_g["original_score"] = max(0,(floatval($gradeable['score'])+floatval($autograding_score)));
+	$this_g["actual_score"] = $this_g["original_score"];
```
Comment (MasterOdin) [inline TAGradingServer/account/submit/admin-grade-summaries.php]: please fix the indentation.

File: TAGradingServer/toolbox/late-day-calculation.php [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+     * @param $userIds Array([String]). The user ids so query for. An empty Array will fetch all users.
+     */
+    function __construct()
+    {
+        //Query database and parse queries
+        $this -> submissions = $this -> get_student_submissions();
+        $this -> latedays = $this -> get_student_lateday_updates();
+        $this -> students = $this -> parse_students($this -> submissions, $this -> latedays);
+        //Calculate lateday usages for all students for all assignments queried
+        $this -> all_latedays = $this -> calculate_student_lateday_usage($this -> students);
```
Comment (MasterOdin) [inline TAGradingServer/toolbox/late-day-calculation.php]: Please don't have a space on either side of `->`.

File: TAGradingServer/toolbox/late-day-calculation.php [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+                    ON
+                      base.user_id = details.user_id
+                      AND base.g_id = details.g_id
+                    ) 
+                      AS submissions 
+                      FULL OUTER JOIN 
+                        late_day_exceptions AS lde 
+                      ON submissions.g_id = lde.g_id 
+                      AND submissions.user_id = lde.user_id";
+        array_push($params, __SUBMISSION_GRACE_PERIOD_SECONDS__);
```
Comment (MasterOdin) [inline TAGradingServer/toolbox/late-day-calculation.php]: `$params[] = __SUBMISSION_GRACE_PERIOD_SECONDS__`

File: TAGradingServer/toolbox/late-day-calculation.php [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+        $students = array();
+
+        //For each submission ensure that an entry exists for that user and append the submission to their list of
+        //submissions.
+        for ($i = 0; $i < count($submissions); $i++) {
+
+            $curr_student = $submissions[$i]['user_id'];
+
+            if (array_key_exists($curr_student, $students)) {
+                array_push($students[$curr_student]['submissions'], $submissions[$i]);
```
Comment (MasterOdin) [inline TAGradingServer/toolbox/late-day-calculation.php]: `$students[$curr_student]['submissions'][] = $submissions[$i];`

**Comment by MasterOdin** [comment]
Asked for some explanation on this:
1) General info on namespaces: https://code.tutsplus.com/tutorials/namespacing-in-php--net-27203
There's then a simple [autoloader](https://github.com/Submitty/Submitty/blob/funky_late_days/TAGradingServer/lib/AutoLoader.php) where that any classes that are referenced via namespaces (by just referencing the class in the code) will auto require the file. This makes things act very similar to how Java works with classes of a project.

Right now, the autoloader adds classes just from TAGradingServer/lib and TAGradingServer/models, so any new class you write should either be in one of those two files or you should add another directory to autoload in functions.php.

**Comment by JerameyATyler** [comment]
@MasterOdin can you put your concerns into a new issue and accept the pull?

**Pull request merged**